apps/studio/src/server/modules/audit/audit.service.ts

```diff
@@ -0,0 +1,181 @@
+import type {
+  AuditLogEvent,
+  Blob,
+  DB,
+  Resource,
+  ResourcePermission,
+  Transaction,
+  User,
+  VerificationToken,
+} from "../database"
+
+type FullResource = Resource & (Blob | undefined)
+
+interface CreateDelta {
+  before: null
+  after: FullResource
+}
+interface DeleteDelta {
+  before: FullResource
+  after: null
+}
+interface UpdateDelta {
+  before: FullResource
+  after: FullResource
+}
+
+interface ResourceEventLogProps {
+  eventType: Extract<
+    AuditLogEvent,
+    "ResourceCreate" | "ResourceUpdate" | "ResourceDelete" | "ResourceMove"
+  >
+  delta: CreateDelta | DeleteDelta | UpdateDelta
+  by: User
+  ip?: string
+}
+
+// NOTE: Type to force every logger method to have a tx
+type AuditLogger<T> = (tx: Transaction<DB>, props: T) => Promise<void>
+
+export const logResourceEvent: AuditLogger<ResourceEventLogProps> = async (
+  tx,
+  { eventType, delta, by, ip },
+) => {
+  await tx
+    .insertInto("AuditLog")
+    .values({ eventType, delta, userId: by.id, ipAddress: ip })
+    .execute()
+}
+
+interface LoginDelta {
+  before: VerificationToken
+  after: VerificationToken
+}
+
+// NOTE: logout just calls `session.destroy` and we only have
+// `userId` or sgid info in session.
```

Comment:
question(non-blocking): wanted to get confirmation, we won't be logging session expiries yeah?

Reply:
session expiry calls `logout` isn't it?

Reply:
Hmm I think not really. The session is set using a cookie with an expiry date, so when the cookie expires it is automatically removed by the browser, and Studio treats the user as if they have never logged in before, so `logout` is not called.

Reply:
eh yea you're right, i m mistaken!